File: apps/api/app/services/profile_token.rb
# Phase 3.9 — encode/decode shareable profile tokens.
#
# `/r/:slug?p=<token>` lets anyone with the URL pre-filter a menu to
# the encoder's profile without signing in. The TypeScript side at
# packages/filter-engine/src/profile-token.ts is the canonical
# implementation; this Ruby side must produce + accept byte-identical
# tokens.
#
# Format: base64url(JSON.generate({ v:, ai:, at:, s:, exp: })). Short
# keys keep URLs reasonable; `v` lets the schema evolve later.
#
# Legal remediation E6 — v2 adds `exp` (Unix-seconds expiry). A shared
# link carries the sharer's avoid-lists + strictness (dietary data), so
# it must not live forever; decode rejects an expired token. v1 tokens
# (no expiry) are no longer accepted. See the TS module's header for why
# the token is not cryptographically signed (it's minted client-side).

class ProfileToken
  VERSION = 2
  STRICTNESSES = %w[relaxed balanced strict].freeze
  # Default lifetime of a shared link: 30 days.
  TTL_SECONDS = 30 * 24 * 60 * 60

  # The TS header calls strict structural validation one of the two real
  # protections here, since the token is minted client-side and signing
  # would be theater. That validation stopped at "an array of strings",
  # which let `ai: ["peanut"]` through — and because the avoid lists are
  # compared in Ruby by array intersection, a junk id matches no dish and
  # is indistinguishable from an empty filter. The person who followed a
  # shared link would be handed a menu that says it is filtered to the
  # sharer's profile and is not filtered at all. Refusing the token says
  # so; the items endpoint already turns this into a 422.
  UUID = /\A\h{8}-\h{4}-\h{4}-\h{4}-\h{12}\z/

  Decoded = Struct.new(:avoid_ingredient_ids, :avoid_tag_ids, :strictness, keyword_init: true)

  class InvalidTokenError < StandardError; end

  class << self
    # `expires_at` (Unix seconds) overrides the default TTL — used by the
    # cross-language parity spec so the encoded bytes are deterministic.
    def encode(avoid_ingredient_ids:, avoid_tag_ids:, strictness:, expires_at: nil)
      payload = {
        v:   VERSION,
        ai:  Array(avoid_ingredient_ids),
        at:  Array(avoid_tag_ids),
        s:   strictness,
        exp: expires_at || (Time.now.to_i + TTL_SECONDS)
      }
      Base64.urlsafe_encode64(JSON.generate(payload), padding: false)
    end

    def decode(token, now: Time.now.to_i)
      raise InvalidTokenError, "empty" if token.to_s.empty?

      json = begin
        Base64.urlsafe_decode64(pad_for_decode(token))
      rescue ArgumentError
        raise InvalidTokenError, "not base64url"
      end

      payload = begin
        JSON.parse(json)
      rescue JSON::ParserError
        raise InvalidTokenError, "not JSON"
      end

      raise InvalidTokenError, "not an object" unless payload.is_a?(Hash)
      raise InvalidTokenError, "unsupported version: #{payload['v']}" unless payload["v"] == VERSION

      ai  = payload["ai"]
      at  = payload["at"]
      s   = payload["s"]
      exp = payload["exp"]

      raise InvalidTokenError, "ai must be an array of ingredient ids" unless id_array?(ai)
      raise InvalidTokenError, "at must be an array of tag ids" unless id_array?(at)
      raise InvalidTokenError, "s must be one of #{STRICTNESSES.join('|')}" unless STRICTNESSES.include?(s)
      raise InvalidTokenError, "exp must be a number" unless exp.is_a?(Numeric)
      raise InvalidTokenError, "expired" if exp <= now

      Decoded.new(avoid_ingredient_ids: ai, avoid_tag_ids: at, strictness: s)
    end

    private

    def id_array?(value)
      value.is_a?(Array) && value.all? { |id| id.is_a?(String) && id.match?(UUID) }
    end

    # `Base64.urlsafe_decode64` requires padding; the token format
    # strips it. Add it back so the decode succeeds.
    def pad_for_decode(token)
      pad = (4 - (token.length % 4)) % 4
      token + ("=" * pad)
    end
  end
end

File: packages/filter-engine/src/profile-token.ts
/**
 * Phase 3.9 — encode shareable profile tokens.
 *
 * `/r/:slug?p=<token>` lets anyone with the URL pre-filter a menu to
 * the encoder's profile without signing in. Both web + mobile mint
 * tokens with `encodeProfileToken` and pass them through opaquely;
 * only Rails decodes, via `?profile_token=` on the items endpoint
 * (`ProfileToken.decode` → `Menus::Filter.from_token`).
 *
 * There is deliberately no TS decoder. Nothing on a client ever needs
 * to read a token back, and the encoding is pinned across languages
 * where it matters: `apps/api/spec/services/profile_token_spec.rb`
 * asserts against a byte-for-byte token literal produced by this
 * encoder, so a change here fails the Ruby suite.
 *
 * Token shape: base64url of JSON `{ v: 2, ai, at, s, exp }`. Short keys
 * keep URLs reasonable; the `v` lets us evolve the schema later
 * without breaking old links.
 *
 * Legal remediation E6 — v2 adds `exp` (a Unix-seconds expiry). A
 * shared link carries the sharer's avoid-lists + strictness (dietary
 * data), so it must not live forever; decoders reject an expired token.
 * v1 tokens (no expiry) are intentionally no longer accepted.
 *
 * Note on signing: these tokens are minted client-side (the Share
 * button), so an HMAC secret would have to ship in the browser bundle
 * and could be forged — signing here would be theater. The token grants
 * no privilege (it only carries the sharer's own filter), so tamper-
 * resistance has no security value; strict structural validation +
 * expiry are the real protections. True signing would require minting
 * the token server-side, a larger change deferred for now.
 */

import type { Strictness } from './index';

export const PROFILE_TOKEN_VERSION = 2;

/** Default lifetime of a shared link: 30 days (in seconds). */
export const PROFILE_TOKEN_TTL_SECONDS = 30 * 24 * 60 * 60;

interface TokenPayload {
  v: number;
  ai: string[];
  at: string[];
  s: Strictness;
  exp: number;
}

export interface EncodeProfileTokenOptions {
  /** Absolute expiry, Unix seconds. Overrides the default TTL (tests/parity). */
  expiresAt?: number;
  /** "Now" in Unix seconds. Defaults to the real clock (tests). */
  nowSeconds?: number;
}

export interface ShareableProfile {
  /**
   * Ingredient and tag **ids**, as UUIDs — the values the server handed
   * back in the filter, which is where both callers already get them.
   * Rails refuses a token whose entries are not ids: the avoid lists are
   * compared by array intersection, so an id that is not an id matches
   * no dish and is indistinguishable from no filter at all, which would
   * show whoever opened the link an unfiltered menu labelled as the
   * sharer's.
   */
  avoid_ingredient_ids: string[];
  avoid_tag_ids: string[];
  strictness: Strictness;
}

export function encodeProfileToken(
  profile: ShareableProfile,
  opts: EncodeProfileTokenOptions = {},
): string {
  const now = opts.nowSeconds ?? Math.floor(Date.now() / 1000);
  const exp = opts.expiresAt ?? now + PROFILE_TOKEN_TTL_SECONDS;
  const payload: TokenPayload = {
    v: PROFILE_TOKEN_VERSION,
    ai: profile.avoid_ingredient_ids,
    at: profile.avoid_tag_ids,
    s: profile.strictness,
    exp,
  };
  return base64UrlEncode(JSON.stringify(payload));
}

// ─── base64url helper (works in both Node and browser) ─────────────

function base64UrlEncode(input: string): string {
  const bytes = utf8Encode(input);
  let b64: string;
  if (typeof Buffer !== 'undefined' && typeof Buffer.from === 'function') {
    b64 = Buffer.from(bytes).toString('base64');
  } else {
    let binary = '';
    for (const byte of bytes) binary += String.fromCharCode(byte);
    b64 = btoa(binary);
  }
  return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

function utf8Encode(s: string): Uint8Array {
  if (typeof TextEncoder !== 'undefined') return new TextEncoder().encode(s);
  // Fallback (very old runtimes — Node < 11). Should never hit in practice.
  const out: number[] = [];
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c < 0x80) out.push(c);
    else if (c < 0x800) out.push(0xc0 | (c >> 6), 0x80 | (c & 0x3f));
    else out.push(0xe0 | (c >> 12), 0x80 | ((c >> 6) & 0x3f), 0x80 | (c & 0x3f));
  }
  return new Uint8Array(out);
}
